// @cliqz/adblocker/dist/cjs/adblocker.cjs
'use strict';

var engine = require('./src/engine/engine.cjs');
var reverseIndex = require('./src/engine/reverse-index.cjs');
var request = require('./src/request.cjs');
var cosmetic = require('./src/filters/cosmetic.cjs');
var network = require('./src/filters/network.cjs');
var lists = require('./src/lists.cjs');
var fetch = require('./src/fetch.cjs');
var utils = require('./src/utils.cjs');
var encoding = require('./src/encoding.cjs');
var config = require('./src/config.cjs');
var resources = require('./src/resources.cjs');
var htmlFiltering = require('./src/html-filtering.cjs');



exports.ENGINE_VERSION = engine.ENGINE_VERSION;
exports.FiltersEngine = engine.default;
exports.ReverseIndex = reverseIndex.default;
exports.Request = request.default;
exports.getHostnameHashesFromLabelsBackward = request.getHostnameHashesFromLabelsBackward;
exports.makeRequest = request.makeRequest;
exports.CosmeticFilter = cosmetic.default;
exports.NetworkFilter = network.default;
exports.detectFilterType = lists.detectFilterType;
exports.f = lists.f;
exports.generateDiff = lists.generateDiff;
exports.getLinesWithFilters = lists.getLinesWithFilters;
exports.mergeDiffs = lists.mergeDiffs;
exports.parseFilter = lists.parseFilter;
exports.parseFilters = lists.parseFilters;
exports.adsAndTrackingLists = fetch.adsAndTrackingLists;
exports.adsLists = fetch.adsLists;
exports.fetchLists = fetch.fetchLists;
exports.fetchResources = fetch.fetchResources;
exports.fetchWithRetry = fetch.fetchWithRetry;
exports.fullLists = fetch.fullLists;
exports.hasUnicode = utils.hasUnicode;
exports.tokenize = utils.tokenizeNoSkip;
exports.isUTF8 = encoding.isUTF8;
exports.Config = config;
exports.Resources = resources;
exports.StreamingHtmlFilter = htmlFiltering.default;
