// @cliqz/adblocker/dist/cjs/src/compression.cjs
'use strict';

var smaz = require('@remusao/smaz');
var cosmeticSelector = require('./codebooks/cosmetic-selector.cjs');
var networkCsp = require('./codebooks/network-csp.cjs');
var networkFilter = require('./codebooks/network-filter.cjs');
var networkHostname = require('./codebooks/network-hostname.cjs');
var networkRedirect = require('./codebooks/network-redirect.cjs');
var rawNetwork = require('./codebooks/raw-network.cjs');
var rawCosmetic = require('./codebooks/raw-cosmetic.cjs');

/*!
 * Copyright (c) 2017-present Cliqz GmbH. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
class Compression {
    constructor() {
        this.cosmeticSelector = new smaz.Smaz(cosmeticSelector);
        this.networkCSP = new smaz.Smaz(networkCsp);
        this.networkRedirect = new smaz.Smaz(networkRedirect);
        this.networkHostname = new smaz.Smaz(networkHostname);
        this.networkFilter = new smaz.Smaz(networkFilter);
        this.networkRaw = new smaz.Smaz(rawNetwork);
        this.cosmeticRaw = new smaz.Smaz(rawCosmetic);
    }
}

module.exports = Compression;
