// @cliqz/adblocker/dist/cjs/src/engine/metadata/organizations.cjs
'use strict';

var map = require('../map.cjs');
var dataView = require('../../data-view.cjs');
var utils = require('../../utils.cjs');

/*!
 * Copyright (c) 2017-present Cliqz GmbH. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
/**
 * This function takes an object representing an organization from TrackerDB
 * dump and validates its shape. The result is the same object, but strongly
 * typed.
 */
function isValid(organization) {
    if (organization === null) {
        return false;
    }
    if (typeof organization !== 'object') {
        return false;
    }
    const { key, name, description, country, website_url: websiteUrl, privacy_policy_url: privacyPolicyUrl, privacy_contact: privacyContact, ghostery_id: ghosteryId, } = organization;
    if (typeof key !== 'string') {
        return false;
    }
    if (typeof name !== 'string') {
        return false;
    }
    if (description !== null && typeof description !== 'string') {
        return false;
    }
    if (country !== null && typeof country !== 'string') {
        return false;
    }
    if (websiteUrl !== null && typeof websiteUrl !== 'string') {
        return false;
    }
    if (privacyPolicyUrl !== null && typeof privacyPolicyUrl !== 'string') {
        return false;
    }
    if (privacyContact !== null && typeof privacyContact !== 'string') {
        return false;
    }
    if (ghosteryId !== null && typeof ghosteryId !== 'string') {
        return false;
    }
    return true;
}
function getKey(organization) {
    return utils.fastHash(organization.key);
}
function getSerializedSize(organization) {
    return (dataView.sizeOfUTF8(organization.key) +
        dataView.sizeOfUTF8(organization.name) +
        dataView.sizeOfUTF8(organization.description || '') +
        dataView.sizeOfUTF8(organization.website_url || '') +
        dataView.sizeOfUTF8(organization.country || '') +
        dataView.sizeOfUTF8(organization.privacy_policy_url || '') +
        dataView.sizeOfUTF8(organization.privacy_contact || '') +
        dataView.sizeOfUTF8(organization.ghostery_id || ''));
}
function serialize(organization, view) {
    view.pushUTF8(organization.key);
    view.pushUTF8(organization.name);
    view.pushUTF8(organization.description || '');
    view.pushUTF8(organization.website_url || '');
    view.pushUTF8(organization.country || '');
    view.pushUTF8(organization.privacy_policy_url || '');
    view.pushUTF8(organization.privacy_contact || '');
    view.pushUTF8(organization.ghostery_id || '');
}
function deserialize(view) {
    return {
        key: view.getUTF8(),
        name: view.getUTF8(),
        description: view.getUTF8() || null,
        website_url: view.getUTF8() || null,
        country: view.getUTF8() || null,
        privacy_policy_url: view.getUTF8() || null,
        privacy_contact: view.getUTF8() || null,
        ghostery_id: view.getUTF8() || null,
    };
}
function createMap(organizations) {
    return new map.CompactMap({
        getSerializedSize,
        getKeys: (organization) => [getKey(organization)],
        serialize,
        deserialize,
        values: organizations,
    });
}

exports.createMap = createMap;
exports.deserialize = deserialize;
exports.getKey = getKey;
exports.getSerializedSize = getSerializedSize;
exports.isValid = isValid;
exports.serialize = serialize;
